nengo/builder/signal.py

```diff
@@ -291,25 +325,29 @@ def __str__(self):
         return sio.getvalue()
 
     def init(self, signal):
-        """Set up a permanent mapping from signal -> ndarray."""
+        """Set up a permanent mapping from signal -> data."""
         if signal in self:
             raise SignalError("Cannot add signal twice")
 
-        x = signal.initial_value
+        assert isinstance(signal, Signal)
+        data = signal.initial_value
+        if isinstance(data, SparseMatrix):
+            data = data.allocate()
+
         if signal.is_view:
             if signal.base not in self:
                 self.init(signal.base)
 
             # get a view onto the base data
             view = np.ndarray(
-                shape=x.shape, strides=x.strides, offset=signal.offset,
-                dtype=x.dtype, buffer=self[signal.base].data)
-            assert np.array_equal(view, x)
+                shape=data.shape, strides=data.strides, offset=signal.offset,
+                dtype=data.dtype, buffer=self[signal.base].data)
+            assert np.array_equal(view, data)
             view.setflags(write=not signal.readonly)
             dict.__setitem__(self, signal, view)
         else:
-            x = x.view() if signal.readonly else x.copy()
-            dict.__setitem__(self, signal, x)
+            data = data if signal.readonly else data.copy()
```

Comment:
So there is a slight difference now that we don't return a view on a readonly array (in the dense case). I don't think this is a big deal, because I don't think we ever change the `shape` of that, or set flags, or things like that. Just wanted to point it out in case anyone is aware of if/how it might affect backends.

Reply:
It was always a view of a view anyway because the base `_initial_value` is always already a readonly view. I tried to make that more clear in `Signal.__init__` through new variable names and comments.

A view of a readonly view is still readonly, there's no practical difference that I can tell.

Reply:
I guess the thing you can't prevent is modifying the flags of the thing that's returned, and so if you modify the flags you get from this then you modify it for all consumers of it, is that the distinction you're pointing out? It's true but if an op is doing something like that to a numpy array that is supposed to be readonly they should just not mark it as readonly.

Reply:
Yeah, that would be the distinction. Even if a view is readonly (i.e. has the `write=False` flag), you can still change the shape of that view (by modifying `.shape`), or modify flags, which would change it for all consumers as you say. But I agree, ops shouldn't be doing these things to readonly arrays.